src/main/java/com/nvidia/spark/rapids/jni/kudo/KudoTableMerger.java

```diff
@@ -0,0 +1,332 @@
+    this.columnOffsets = columnOffsets;
+    this.buffer = buffer;
+    this.colViewInfoList = new ArrayList<>(columnOffsets.size());
+  }
+
+  @Override
+  protected KudoHostMergeResult doVisitTopSchema(Schema schema, List<Void> children) {
+    return new KudoHostMergeResult(schema, buffer, colViewInfoList);
+  }
+
+  @Override
+  protected Void doVisitStruct(Schema structType, List<Void> children) {
+    ColumnOffsetInfo offsetInfo = getCurColumnOffsets();
+    int nullCount = deserializeValidityBuffer(offsetInfo);
+    int totalRowCount = getTotalRowCount();
+    colViewInfoList.add(new ColumnViewInfo(structType.getType(),
+        offsetInfo, nullCount, totalRowCount));
+    return null;
+  }
+
+  @Override
+  protected Void doPreVisitList(Schema listType) {
+    ColumnOffsetInfo offsetInfo = getCurColumnOffsets();
+    int nullCount = deserializeValidityBuffer(offsetInfo);
+    int totalRowCount = getTotalRowCount();
+    deserializeOffsetBuffer(offsetInfo);
+
+    colViewInfoList.add(new ColumnViewInfo(listType.getType(),
+        offsetInfo, nullCount, totalRowCount));
+    return null;
+  }
+
+  @Override
+  protected Void doVisitList(Schema listType, Void preVisitResult, Void childResult) {
+    return null;
+  }
+
+  @Override
+  protected Void doVisit(Schema primitiveType) {
+    ColumnOffsetInfo offsetInfo = getCurColumnOffsets();
+    int nullCount = deserializeValidityBuffer(offsetInfo);
+    int totalRowCount = getTotalRowCount();
+    if (primitiveType.getType().hasOffsets()) {
+      deserializeOffsetBuffer(offsetInfo);
+      deserializeDataBuffer(offsetInfo, OptionalInt.empty());
+    } else {
+      deserializeDataBuffer(offsetInfo, OptionalInt.of(primitiveType.getType().getSizeInBytes()));
+    }
+
+    colViewInfoList.add(new ColumnViewInfo(primitiveType.getType(),
+        offsetInfo, nullCount, totalRowCount));
+
+    return null;
+  }
+
+  private int deserializeValidityBuffer(ColumnOffsetInfo curColOffset) {
+    if (curColOffset.getValidity() != INVALID_OFFSET) {
+      long offset = curColOffset.getValidity();
+      long validityBufferSize = padFor64byteAlignment(getValidityLengthInBytes(getTotalRowCount()));
+      try (HostMemoryBuffer validityBuffer = buffer.slice(offset, validityBufferSize)) {
+        int nullCountTotal = 0;
+        int startRow = 0;
+        for (int tableIdx = 0; tableIdx < getTableSize(); tableIdx += 1) {
+          SliceInfo sliceInfo = sliceInfoOf(tableIdx);
+          long validityOffset = validifyBufferOffset(tableIdx);
+          if (validityOffset != INVALID_OFFSET) {
+            nullCountTotal += copyValidityBuffer(validityBuffer, startRow,
+                memoryBufferOf(tableIdx), toIntExact(validityOffset),
+                sliceInfo);
+          } else {
+            appendAllValid(validityBuffer, startRow, sliceInfo.getRowCount());
+          }
+
+          startRow += sliceInfo.getRowCount();
+        }
+        return toIntExact(nullCountTotal);
```

Comment:
nullCountTotal is already an int, so this is wasteful.  There should already be a row count overflow check elsewhere (and we cannot overflow nullcounts without also overflowing row counts), so seems like this should simply return nullCountTotal directly.